chain/src/chain.rs

```diff
@@ -281,52 +244,223 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
             if let Some(old_hash) = self.shared.block_hash(number) {
                 if old_hash == hash {
+                    verified = true;
                     break;
                 }
+
                 let old_block = self.shared.block(&old_hash).unwrap();
 
-                self.shared.store().delete_block_hash(batch, number);
-                self.shared.store().delete_block_number(batch, &old_hash);
-                self.shared
-                    .store()
-                    .delete_transaction_address(batch, &old_block.commit_transactions());
-                old_cumulative_blks.push(old_block);
+                old_blocks.push(old_block);
             }
 
-            let new_block = self.shared.block(&hash).unwrap();
+            if !verified {
+                let ext = self.shared.block_ext(&hash).unwrap();
+                if let Some(x) = ext.valid {
+                    if x {
+                        verified = true;
+                    } else {
+                        break;
+                    }
+                } else {
+                    exts.push((hash.clone(), ext));
+                }
+            }
 
-            self.shared.store().insert_block_hash(batch, number, &hash);
-            self.shared
-                .store()
-                .insert_block_number(batch, &hash, number);
-            self.shared.store().insert_transaction_address(
-                batch,
-                &hash,
-                &new_block.commit_transactions(),
-            );
+            let new_block = self.shared.block(&hash).unwrap();
 
             hash = new_block.header().parent_hash().clone();
             number -= 1;
 
-            new_cumulative_blks.push(new_block);
+            new_blocks.push(new_block);
+        }
+
+        (verified, old_blocks, new_blocks)
+    }
+
+    // we found new best_block total_difficulty > old_chain.total_difficulty
+    fn reconcile_main_chain(
+        &self,
+        batch: &mut Batch,
+        tip_number: BlockNumber,
+        block: &Block,
+        ext: BlockExt,
+        chain_state: &ChainState,
+    ) -> Result<(TxoSetDiff, Vec<Block>, Vec<Block>), SharedError> {
+        let skip_verify = !self.shared.consensus().verification;
+        let mut exts = Vec::new();
+        let (mut verified, old_blocks, mut new_blocks) =
+            self.get_forks(tip_number, block, &mut exts, ext);
+
+        //new valid block len
+        let mut new_len = 0;
+
+        let mut old_inputs = FnvHashSet::default();
+        let mut old_outputs = FnvHashSet::default();
+        let mut new_inputs = FnvHashSet::default();
+        let mut new_outputs = FnvHashMap::default();
+
+        if verified {
+            let new_blocks_iter = new_blocks.iter().rev();
+            let old_blocks_iter = old_blocks.iter().rev();
+            let new_blocks_len = new_blocks.len();
+
+            let verified_len = new_blocks_len - exts.len();
+
+            for b in old_blocks_iter.skip(verified_len) {
```

Comment:
Why skip `verified_len` old blocks here?

Reply:
Because they have been verified, we do not need verify it again.

Reply:
But doesn't all old blocks have been verified? Since these old blocks are in the previous main chain. Also `verified_len` is counted based on new blocks and the new blocks pending verification, I don't see a relationship between old blocks here.